Add support for sending POST requests as queries

OverloadUT opened the pull request:
POST queries require a client identifier

phillipj commented on lib/api.js:
```diff
@@ -146,6 +163,52 @@ PlexAPI.prototype._request = function _request(relativeUrl) {
     return deferred.promise;
 };
 
+PlexAPI.prototype._post = function perform(relativeUrl) {
```
`_post()` looks very similar to `_request()`. Maybe `_request()` could be refactored to handle both GET and POST?

phillipj commented on lib/api.js:
```diff
@@ -153,6 +216,11 @@ PlexAPI.prototype._authenticateAndRetry = function _authenticateAndRetry(url, re
+    // Bit of a kludgey way to generate an Identifier before we send off the options to auth.
+    // Right now it doesn't appear that using the same identifier on every request is necessary, but
+    // it's probably good practice to do so.
+    self.getIdentifier();
```
As this ensures an identifier is always generated, should it be moved into the constructor instead? Something like

```js
this.options.identifier = this.options.identifier || uuid.v4();
```
on line #21?

phillipj commented:
Thanks for more contributions @OverloadUT!

A couple of tests covering POST requests + documenting these two methods you have added in the readme would have been nice :smiley:

OverloadUT commented:
I will add those! I had meant to, but totally forgot. I'm fixing the code formatting issues that Travis pointed out as well.

OverloadUT commented:
Alright, I did a big refactoring for this PR. When combining _request and _post, it just didn't make sense to also have perform() there doing 90% of the same stuff _request does. So I moved it so that all three kinds of requests (other than auth) go through _request.

Also added tests and documentation :)

phillipj commented:
Wow, great refactoring @OverloadUT! I'll merge and publish v2.4.0 right away :smiley:

Would you like to be added as a collaborator of this project?

OverloadUT commented:
I'd be happy to be added!

I've got one more PR coming up that will then catch it up to my "xml2js instead of xml2json" fork that I'm using.

phillipj commented:
There you go, welcome onboard :)

What's the benefit of using xml2js compared to xml2json?

OverloadUT commented:
The tl;dr is that I need it to not require a C library.

It's a very specific requirement in my case: I am deploying my code to **AWS Lambda**. The way that works is I need to upload a zip of the project which can run on Amazon's linux distro. The problem is that I'm developing on Windows, so when I ```npm install``` my module and upload, it ends up with a Windows-compiled C library which of course doesn't work in the linux environment. (It also makes the deploy package massively larger, but that doesn't *really* matter)

The way to really fix it on my side is to have an Amazon Linux VM and do my deploys from there. I might set that up one of these days, but for now it's been much easier to simply change two lines of code in node-plex-api.

phillipj commented:
Getting rid of the native addon sounds like a very good idea. Especially
when there's no appearent reason to use native. I've spent a good deal of
hours debugging such modules, not always trivial for non-native-ninjas.
On fre. 3. jul. 2015 at 23.13 Greg Laabs <notifications@github.com> wrote:

> The tl;dr is that I need it to not require a C library.
>
> It's a very specific requirement in my case: I am deploying my code to *AWS
> Lambda*. The way that works is I need to upload a zip of the project
> which can run on Amazon's linux distro. The problem is that I'm developing
> on Windows, so when I npm install my module and upload, it ends up with a
> Windows-compiled C library which of course doesn't work in the linux
> environment. (It also makes the deploy package massively larger, but that
> doesn't *really* matter)
>
> The way to really fix it on my side is to have an Amazon Linux VM and do
> my deploys from there. I might set that up one of these days, but for now
> it's been much easier to simply change two lines of code in node-plex-api.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phillipj/node-plex-api/pull/9#issuecomment-118426058>.
>
